compiler/rustc_metadata/src/rmeta/decoder/cstore_impl.rs

```diff
@@ -493,6 +494,17 @@ pub(in crate::rmeta) fn provide(providers: &mut Providers) {
             tcx.arena
                 .alloc_slice(&CStore::from_tcx(tcx).crate_dependencies_in_postorder(LOCAL_CRATE))
         },
+        is_in_upstream_dylib: |tcx, cnum| {
```

Comment:
Maybe call this `may_come_from_upstream_dylib`? Also note that when compiling an rlib `Linkage::NotIncluded` will be used and as such the current impl may return `false` incorrectly. Given that rlib is the most common crate type with an overwhelming margin, maybe just always return `true`?

Reply:
Yeah I guess this is currently wrong. I want this to be accurate so that rlib to rlib references avoids calling the MIR shims.

Reply:
That is unavoidable. Any rlib dep can be included in a foreign dylib behind the back of the local rlib. There is nothing that can prevent this.

Reply:
I guess rlib code generation just has to suffer then :(